File: 프로그래머스/1/250121. ［PCCE 기출문제］ 10번 ／ 데이터 분석/［PCCE 기출문제］ 10번 ／ 데이터 분석.java
import java.util.*;

class Solution {
    static String curSortBy;
    
    static class Data implements Comparable<Data>{
        int code;
        int date;
        int maximum;
        int remain;
        
        Data(int code, int date, int maximum, int remain) {
            this.code = code;
            this.date = date;
            this.maximum = maximum;
            this.remain = remain;
        }
        
        @Override
        public int compareTo(Data d) {
            if (curSortBy.equals("code")) {
                return this.code - d.code;
            } else if (curSortBy.equals("date")) {
                return this.date - d.date;
            } else if (curSortBy.equals("maximum")) {
                return this.maximum - d.maximum;
            } else if (curSortBy.equals("remain")) {
                return this.remain - d.remain;
            }
            return 0;
        }
    }
    public int[][] solution(int[][] data, String ext, int val_ext, String sort_by) {
        
        List<Data> datas = new ArrayList<>();
        for (int[] d : data) {
            Data newData = new Data(d[0],d[1],d[2],d[3]);
            if (ext.equals("code")) {
                if (d[0] >= val_ext) continue;
            } else if (ext.equals("date")) {
                if (d[1] >= val_ext) continue;
            } else if (ext.equals("maximum")) {
                if (d[2] >= val_ext) continue;
            } else if (ext.equals("remain")) {
                if (d[3] >= val_ext) continue;
            }
            datas.add(newData);
        }
        curSortBy = sort_by;
        
        Collections.sort(datas);
        int[][] answer = new int[datas.size()][4];
        for (int i = 0; i < datas.size(); i++) {
            answer[i][0] = datas.get(i).code;
            answer[i][1] = datas.get(i).date;
            answer[i][2] = datas.get(i).maximum;
            answer[i][3] = datas.get(i).remain;
        }
        
        return answer;
    }
}

File: 프로그래머스/1/250121. ［PCCE 기출문제］ 10번 ／ 데이터 분석/［PCCE 기출문제］ 10번 ／ 데이터 분석.py
def solution(data, ext, val_ext, sort_by):
    answer = []
    
    def get_year(data):
        return data // 10000
    
    def get_month(data):
        return (data % 10000) // 100
    
    def get_date(data):
        return (data % 100)
    
    def check(data, index):
        if data[index] < val_ext:
            return True
        return False
    
    def check_date(data, val_year, val_month, val_date):
        cur_year = get_year(data)
        cur_month = get_month(data)
        cur_date = get_date(data)
        if cur_year <= val_year:
            if cur_month < val_month:
                return True
            elif cur_month == val_month:
                if cur_date < val_date:
                    return True
        return False
    
    if ext == 'code':
        for d in data:
            if check(d, 0):
                answer.append(d)
        
    elif ext == 'date':
        val_year = get_year(val_ext)
        val_month = get_month(val_ext)
        val_date = get_date(val_ext)
        
        for d in data:
            if check_date(d[1], val_year, val_month, val_date):
                answer.append(d)
        
    elif ext == 'maximum':
        for d in data:
            if check(d, 2):
                answer.append(d)
        
    elif ext == 'remain':
        for d in data:
            if check(d, 3):
                answer.append(d)
            
    key_index = 0
    if sort_by == 'code':
        key_index = 0
    elif sort_by == 'date':
        key_index = 1
    elif sort_by == 'maximum':
        key_index = 2
    elif sort_by == 'remain':
        key_index = 3
        
    answer.sort(key = lambda x: x[key_index])
        
    return answer